x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/hash/Hash.java

```diff
@@ -0,0 +1,153 @@
+    ) {
+        super(source, List.of(alg, input));
+        this.alg = alg;
+        this.input = input;
+    }
+
+    private Hash(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(alg);
+        out.writeNamedWriteable(input);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(alg, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isString(input, sourceText(), SECOND);
+    }
+
+    @Override
+    public boolean foldable() {
+        return alg.foldable() && input.foldable();
+    }
+
+    @Evaluator(warnExceptions = NoSuchAlgorithmException.class)
+    static BytesRef process(@Fixed(includeInToString = false, build = true) BreakingBytesRefBuilder scratch, BytesRef alg, BytesRef input)
+        throws NoSuchAlgorithmException {
+        return hash(scratch, MessageDigest.getInstance(alg.utf8ToString()), input);
+    }
+
+    @Evaluator(extraName = "Constant")
+    static BytesRef processConstant(
+        @Fixed(includeInToString = false, build = true) BreakingBytesRefBuilder scratch,
+        @Fixed(build = true) MessageDigest alg,
+        BytesRef input
+    ) {
+        return hash(scratch, alg, input);
+    }
+
+    private static BytesRef hash(BreakingBytesRefBuilder scratch, MessageDigest alg, BytesRef input) {
+        alg.reset();
+        alg.update(input.bytes, input.offset, input.length);
+        var digest = alg.digest();
+        scratch.clear();
+        scratch.grow(digest.length * 2);
+        scratch.append(new BytesRef(HexFormat.of().formatHex(digest)));
+        return scratch.bytesRefView();
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        if (alg.foldable() && alg.dataType() == DataType.KEYWORD) {
+            try {
+                var md = MessageDigest.getInstance(((BytesRef) alg.fold()).utf8ToString());
+                return new HashConstantEvaluator.Factory(
+                    source(),
+                    context -> new BreakingBytesRefBuilder(context.breaker(), "hash"),
+                    context -> md,
+                    toEvaluator.apply(input)
+                );
+            } catch (NoSuchAlgorithmException e) {
+                throw new IllegalArgumentException(e);
```

Comment:
I wonder what should be the right way to handle this failure?
Should it be done with 
```
Warnings.createWarnings(
              driverContext.warningsMode(),
              source.source().getLineNumber(),
              source.source().getColumnNumber(),
              source.text()
          ).registerException(e);
```
If so, is there a way to access `driverContext` at this stage?

Reply:
Not here, no. We get a `Warnings` as part of building the evaluator. That's important because then we only get at most one per evaluator.